### PR TITLE
Add support for multiuser client in python

### LINKED ISSUE #21: Add a MultiUserClient for Python
The `MutliUserClient` is a useful abstraction in the node sdk for applications that manage multiple users calling into Redactive, we should add it to the python sdk as well.

### PR DESCRIPTION
**Pull request by IsisChameleon**
Implement Feature request #21

### CODE REVIEW

File: sdks/python/src/redactive/multi_user_client.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+
+class MultiUserClient:
+    def __init__(
+        self,
+        api_key: str,
+        callback_uri: str,
+        read_user_data: Callable[[str], Awaitable[UserData]],
+        write_user_data: Callable[[str, UserData | None], Awaitable[None]],
+        config: MultiUserClientConfig,
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]: Two suggestions with the `config` parameter: 
- I don't reckon it should be defined by a `MultiUserClientConfig` class & instead be a series of keyword-arguments; e.g. `MultiUserClient(..., auth_base_url=..., ...)`
- These parameters should be optional, similar to how it is implemented in the [Node SDK](https://github.com/redactive-ai/redactive/blob/ad96a8ced8a750b80684381b7b0794ea3be753fb/sdks/node/src/multiUserClient.ts#L45)

File: sdks/python/src/redactive/multi_user_client.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        super().__init__(f"No valid Redactive session for user '{user_id}'")
+
+
+class MultiUserClient:
+    def __init__(
+        self,
+        api_key: str,
+        callback_uri: str,
+        read_user_data: Callable[[str], Awaitable[UserData]],
+        write_user_data: Callable[[str, UserData | None], Awaitable[None]],
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]: Thoughts on using annotations for the type hinting? e.g. 
```
read_user_data: Callable[[Annotated[str, 'user_id']], Awaitable[UserData]],
write_user_data: Callable[[Annotated[str, 'user_id'], UserData | None], Awaitable[None]],
```

File: sdks/python/src/redactive/multi_user_client.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        config: MultiUserClientConfig,
+    ) -> None:
+        """Redactive client handling multiple users authentication and access to the Redactive Search service.
+
+        :param api_key: Redactive API key.
+        :type api_key: str
+        :param callback_uri: The URI to redirect to after initiating the connection. Defaults to an empty string.
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]: I do not believe this "Defaults to an empty string."

File: sdks/python/src/redactive/multi_user_client.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        self.auth_client = AuthClient(api_key, base_url=config.auth_base_url)
+        self.search_client = SearchClient(config.grpc_host, config.grpc_port)
+        self.callback_uri = callback_uri
+        self.read_user_data = read_user_data
+        self.write_user_data = write_user_data
+
+    async def get_begin_connection_url(self, user_id: str, provider: str) -> str:
+        state = str(uuid.uuid4())
+        response = await self.auth_client.begin_connection(provider, self.callback_uri, code_param_alias=state)
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]: There's an issue with the Python SDK's AuthClient's `begin_connection`. It does not implement `state` parameter as is in the Node SDK [here](https://github.com/redactive-ai/redactive/blob/ad96a8ced8a750b80684381b7b0794ea3be753fb/sdks/node/src/authClient.ts#L35). `code_param_alias` =/= `state`. Can you correct this?

File: sdks/python/src/redactive/multi_user_client.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        user_data.sign_in_state = state
+        await self.write_user_data(user_id, user_data)
+        return response.url
+
+    async def _refresh_user_data(
+        self, user_id: str, refresh_token: str | None = None, sign_in_code: str | None = None
+    ) -> UserData:
+        tokens = await self.auth_client.exchange_tokens(sign_in_code, refresh_token)
+        connections: ListConnectionsResponse = await self.auth_client.list_connections(tokens.idToken)
+        user_data: UserData = UserData(
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]:
```suggestion
        user_data = UserData(
```
I think a type-hint here is unnecessary if the line is creating that type

File: sdks/python/src/redactive/multi_user_client.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            tokens.refreshToken,
+            tokens.idToken,
+            datetime.now(UTC) + timedelta(seconds=tokens.expiresIn - 10),
+            connections.connections,
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]: Could these parameters be keyword-assigned please; e.g. `refresh_token=tokens.refreshToken`

File: sdks/python/src/redactive/multi_user_client.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        user_data = await self.read_user_data(user_id)
+        if not user_data or user_data.sign_in_state != state:
+            return False
+        await self._refresh_user_data(user_id, sign_in_code=sign_in_code)
+        return True
+
+    async def get_user_connections(self, user_id: str) -> list[str]:
+        user_data = await self.read_user_data(user_id)
+        if user_data and user_data.id_token_expiry and user_data.id_token_expiry > datetime.now(UTC):
+            return user_data.connections or []
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]:
```suggestion
            return user_data.connections
```
L18 `UserData.connections` defaults to an empty list

File: sdks/python/src/redactive/multi_user_client.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        if user_data and user_data.refresh_token:
+            user_data = await self._refresh_user_data(user_id, refresh_token=user_data.refresh_token)
+            return user_data.connections or []
```
Comment (MattTimms) [inline sdks/python/src/redactive/multi_user_client.py]:
```suggestion
            return user_data.connections
```
L18 `UserData.connections` defaults to an empty list

File: sdks/python/tests/unit_tests/multi_user_client_tests.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+import pytest
+
+from redactive.auth_client import AuthClient
+from redactive.grpc.v1 import Chunk, RelevantChunk
+from redactive.multi_user_client import MultiUserClient, MultiUserClientConfig, UserData
+from redactive.search_client import SearchClient
+
+
+def mock_read_user_data(user_id) -> UserData:
```
Comment (MattTimms) [inline sdks/python/tests/unit_tests/multi_user_client_tests.py]:
```suggestion
def mock_read_user_data(user_id) -> UserData | None:
```

**Review by MattTimms — COMMENTED** [review]
I've left some comments :)

File: sdks/python/src/redactive/multi_user_client.py (hunk 1 above)
Comment (IsisChameleon) [reply sdks/python/src/redactive/multi_user_client.py]: OK done like here:
    options?: {
      authBaseUrl?: string;
      grpcHost?: string;
      grpcPort?: number;
    }

Comment (MattTimms) [reply sdks/python/src/redactive/multi_user_client.py]: Yeah, so I'm suggesting something like:
```python
def __init__(
        self,
        api_key: str,
        callback_uri: str,
        read_user_data: Callable[[str], Awaitable[UserData]],
        write_user_data: Callable[[str, UserData | None], Awaitable[None]],
        *,
        auth_base_url: str | None = None
        grpc_host: str | None = None
        grpc_port: int | None = None
)
```

File: sdks/python/src/redactive/multi_user_client.py (hunk 4 above)
Comment (IsisChameleon) [reply sdks/python/src/redactive/multi_user_client.py]: Ok!

**Review by MattTimms — APPROVED** [review]

**Pull request merged**